Support auth0 login

franconob opened the pull request:
## Background
The application was missing an important feature: authentication. This PR adds the feature to sign into the app using Auth0 API. The authentication calls Auth0 API endpoint, which indeed makes a request to the local database to grant access to the user. 
## Changes done
Many changes have been done. There are two components that need special attention:
1. **AuthenticationManager**: this component wraps all other components and checks if the user is already authenticated. If not, it redirects to the login page. You can see its implementation in the `app/routes.js` file.
2. **CallbackLoading**: This one is mounted on the /callback path. It's responsible for calling the `authService.handleAuthentication function` and based on the returned value, dispatches the `loginSuccess` or `loginFailed` redux actions. It also renders a simple `<div>Loading...</div>` indicator while the `authService` executes the validation.

Besides those two components strictly related to handling authentication and redirection, this PR implements two different layouts: `app/components/layout/SignedInLayout` and `app/components/layout/SignedOutLayout`, both inheriting from `app/components/layout/BaseLayout`. This decision was taken to keep components not aware of things like `if (authService.isAuthenticated) {...}`, especially on the Header component, which renders the top-bar and shows the name of the logged in user.
## Pending to be done
There are two pending tasks, ~~one of them is to use the [redux-persist](https://github.com/rt2zz/redux-persist) package to save the user session state in the localStorage.~~ The other one is to write tests for the new components.
## Notes
The Auth0 API must be configured. There is a new package installed, called [dovent-webpack](https://github.com/mrsteele/dotenv-webpack) which allows defining a `.env` file. It's similar to the `config` package, but this one only works for Node apps (it depends on the `fs` module). There are 5 settings you must define here, the name of them are quite self-explanatory. There is a `.env.dist` showing those settings.
Another thing to keep in mind is that Auth0 uses our local database to check for a valid user. This means that you have to expose the mongodb port to the internet

Commit:
Support auth0 login

Add two different layouts

Created two layouts, one for signed in users and the other for holding login page

Add auth components

AuthenticationManager handles redirection to /login if the user is not authenticated
CallbackLoading handles callback url and dispatch loginSuccess or loginFalied actions
Logout calls AuthService.logout and dispatches logoutAction
LoginForm and LoginPage are self-explanatory

Add new redux auth actions

Show signed in user's name

Add new authReducer

Auth reducer handles actions like LOGIN_SUCCESS, LOGIN_FAILED and LOGOUT

Add new auth service to interact with auth0

Add new auth routes

Remove App component

This component was replaced by SignedInLayout and SignedOutLayout layout components.

Add password field to user's form and to db schema.

The password is encrypted using bcrypt module

Exclude password field from react

Install bcrypt module

Commit:
Update package-lock.json to the latest version

Commit:
Use prop-types package

Commit:
Fix reducers path in mapStateToProps function

Commit:
Fix path property in SingedInLayout route

Commit:
Use localStorage to save token and expiration date

Commit:
Implement JWT authentication in Express

Commit:
Install dotenv-webpack webpack plugin to manage env aware configuration

Commit:
Mock localStorage in tests

mravinale commented:
great work!, let's add redirection in case when we need to access an internal url and we don't have token @franconob

Commit:
Fix missing database parameter

Commit:
Fix UnauthorizedError when an user visits a protected url

This problem is sovled by redirecting the user to the homepage ("/")
when no token is found. This was also possible with the help of a new
package, redux-persist, which persists the redux state in localStorage
